src/main/java/org/opensearch/knn/quantization/sampler/ReservoirSampler.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.quantization.sampler;
+
+import java.util.BitSet;
+import java.util.Random;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * ReservoirSampler implements the Sampler interface and provides a method for sampling
+ * a specified number of indices from a total number of vectors using the reservoir sampling algorithm.
+ * This algorithm is particularly useful for randomly sampling a subset of data from a larger set
+ * when the total size of the dataset is unknown or very large.
+ */
+final class ReservoirSampler implements Sampler {
+
+    private final Random random;
+
+    /**
+     * Constructs a ReservoirSampler with a new Random instance.
+     */
+    public ReservoirSampler() {
+        this.random = ThreadLocalRandom.current();
```

Comment:
Won't caching this as a member variable defeat the purpose of isolating random to a single thread? 

The sampler is a member variable for quantizer. so there is one instance of this created. If train is called in parallel, will there be a risk of contention here?

Typical use of this will be wherever a new random number needs to be generated in this case on line 67

```
int j = ThreadLocalRandom.current().nextInt(i + 1);
```